lib/ed25519.js

```diff
@@ -0,0 +1,140 @@
+ * Allows Uint8Arrays to be interoperable with node's crypto functions.
+ *
+ * @param {object} options - Options to use.
+ * @param {Buffer} [options.privateKeyBytes] - Required if no seedBytes.
+ * @param {Buffer} [options.seedBytes] - Required if no privateKeyBytes.
+ *
+ * @throws {TypeError} Throws if the supplied buffer is not of the right size
+ *  or not a Uint8Array or Buffer.
+ *
+ * @returns {Buffer} DER private key prefix + key bytes.
+*/
+function privateKeyDerEncode({privateKeyBytes, seedBytes}) {
+  if(!(privateKeyBytes || seedBytes)) {
+    throw new TypeError('`privateKeyBytes` or `seedBytes` is required.');
+  }
+  if(!privateKeyBytes && !(seedBytes instanceof Uint8Array &&
+    seedBytes.length === 32)) {
+    throw new TypeError('`seedBytes` must be a 32 byte Buffer.');
+  }
+  if(!seedBytes && !(privateKeyBytes instanceof Uint8Array &&
+    privateKeyBytes.length === 64)) {
+    throw new TypeError('`privateKeyBytes` must be a 64 byte Buffer.');
+  }
+  let p;
+  if(seedBytes) {
+    p = seedBytes;
+  } else {
+    // extract the first 32 bytes of the 64 byte private key representation
+    p = privateKeyBytes.slice(0, 32);
+  }
+  return Buffer.concat([DER_PRIVATE_KEY_PREFIX, p]);
+}
+
+/**
+ * Takes a Buffer of public key bytes and encodes it in DER-encoded
+ * SubjectPublicKeyInfo (SPKI) format.
+ * Allows Uint8Arrays to be interoperable with node's crypto functions.
+ *
+ * @param {object} options - Options to use.
+ * @param {Uint8Array} options.publicKeyBytes - The keyBytes.
+ *
+ * @throws {TypeError} Throws if the bytes are not Uint8Array or of length 32.
+ *
+ * @returns {Buffer} DER Public key Prefix + key bytes.
+*/
+function publicKeyDerEncode({publicKeyBytes}) {
+  if(!(publicKeyBytes instanceof Uint8Array && publicKeyBytes.length === 32)) {
+    throw new TypeError('`publicKeyBytes` must be a 32 byte Buffer.');
+  }
+  return Buffer.concat([DER_PUBLIC_KEY_PREFIX, publicKeyBytes]);
+}
+
+const api = {
+  /**
+   * Generates a key using a 32 bit Uint8Array.
+   *
+   * @param {Uint8Array} seedBytes - The bytes for the private key.
+   *
+   * @returns {object} The buffer for the public and private key material.
+  */
+  async generateKeyPairFromSeed(seedBytes) {
+    const privateKey = await createPrivateKey({
+      // node is more than happy to create a new private key using a DER
+      key: privateKeyDerEncode({seedBytes}),
+      format: 'der',
+      type: 'pkcs8'
+    });
+    // this expects either a PEM encoded key or a node privateKeyObject
+    const publicKey = await createPublicKey(privateKey);
+    const publicKeyBuffer = publicKey.export(publicKeyEncoding);
+    const publicKeyBytes = getKeyMaterial(publicKeyBuffer);
+    return {
+      publicKey: publicKeyBytes,
+      secretKey: Buffer.concat([seedBytes, publicKeyBytes])
+    };
+  },
+  // generates an ed25519 key using a random seed
+  async generateKeyPair() {
+    return api.generateKeyPairFromSeed(randomBytes(32));
```

Comment:
Note: This will use the synchronous version of `randomBytes`. I was suggesting we use the asynchronous version (which requires promisifying it) -- but we can change to that later if it's really an issue. I won't hold this PR up over the difference -- and we don't have data indicating which would be better in a given system anyway. It isn't likely to have a significant event loop blocking impact (I don't think node.js's PRNG implementation ever blocks for significant periods of time).

Reply:
if you can link me to the promisify version that is fine, but as far I can tell it now returns a promise it looks like.

Reply:
Yes, this is what I was suggesting we do to ensure we don't block the event loop.